File: Alkitab/src/main/java/yuku/alkitab/reminder/widget/ReminderTimePreference.java
package yuku.alkitab.reminder.widget;

import android.content.Context;
import android.support.v7.preference.Preference;
import android.text.format.DateFormat;
import android.util.AttributeSet;
import android.widget.TimePicker;
import yuku.alkitab.debug.R;

import java.util.Locale;

public class ReminderTimePreference extends Preference {
	public ReminderTimePreference(final Context context, final AttributeSet attrs, final int defStyle) {
		super(context, attrs, defStyle);
	}

	public ReminderTimePreference(final Context context, final AttributeSet attrs) {
		super(context, attrs);
	}

	@Override protected void onClick() {
		super.onClick();

		String currentValue = getPersistedString(null);
		final int hour = currentValue == null? 12: Integer.parseInt(currentValue.substring(0, 2));
		final int minute = currentValue == null? 0: Integer.parseInt(currentValue.substring(2, 4));

		final Context context = getContext();
		HackedTimePickerDialog dialog = new HackedTimePickerDialog(context, getTitle(), context.getString(R.string.dr_timepicker_set), context.getString(R.string.dr_timepicker_off), new HackedTimePickerDialog.HackedTimePickerListener() {
			@Override public void onTimeSet(TimePicker view, int hourOfDay, int minute) {
				persistString(String.format(Locale.US, "%02d%02d", hourOfDay, minute));
				notifyChanged();
			}

			@Override public void onTimeOff(TimePicker view) {
				persistString(null);
				notifyChanged();
			}
		}, hour, minute, DateFormat.is24HourFormat(context));
		dialog.show();
	}

	@Override public boolean shouldDisableDependents() {
		return getPersistedString(null) == null;
	}
}

File: Alkitab/src/main/java/yuku/alkitab/reminder/widget/ReminderTimePreference.kt
package yuku.alkitab.reminder.widget

import android.content.Context
import android.os.Build
import android.text.InputType
import android.text.format.DateFormat
import android.util.AttributeSet
import androidx.preference.Preference
import com.afollestad.materialdialogs.MaterialDialog
import yuku.alkitab.debug.R
import java.util.Locale

class ReminderTimePreference @JvmOverloads constructor(context: Context, attrs: AttributeSet, defStyle: Int = 0) : Preference(context, attrs, defStyle) {
    override fun onClick() {
        super.onClick()

        val currentValue = getPersistedString(null)

        val listener = object : HackedTimePickerDialog.HackedTimePickerListener {
            override fun onTimeSet(hourOfDay: Int, minute: Int) {
                persistString(formatForPersistence(hourOfDay, minute))
                notifyChanged()
            }

            override fun onTimeOff() {
                persistString(null)
                notifyChanged()
            }
        }

        val hour = currentValue?.substring(0, 2)?.toIntOrNull() ?: 12
        val minute = currentValue?.substring(2, 4)?.toIntOrNull() ?: 0

        HackedTimePickerDialog(
            context,
            title,
            context.getString(R.string.dr_timepicker_set),
            context.getString(R.string.dr_timepicker_off),
            listener,
            hour,
            minute,
            DateFormat.is24HourFormat(context)
        ).show()
    }

    private fun formatForPersistence(hour: Int, minute: Int) = String.format(Locale.US, "%02d%02d", hour, minute)

    override fun shouldDisableDependents(): Boolean {
        return getPersistedString(null) == null
    }
}
